Move DNS trace latency calculation to BPF

wedaly opened the pull request:
# Move DNS trace latency calculation to BPF

Previously, DNS trace latency was calculated in userspace by inserting and retrieving events into an LRU hashmap. This implicitly assumed that events were ordered ascending by timestamp, but BPF perf ring buffers do not guarantee chronological order (see https://kinvolk.io/blog/2018/02/timing-issues-when-using-bpf-with-virtual-cpus/)

This resulted in missing latency fields in the DNS trace events and caused the TestTraceDns integration test to fail randomly, as reported in https://github.com/inspektor-gadget/inspektor-gadget/issues/1440

To address this problem, move the latency calculation from userspace to the BPF program, with query timestamps stored in a BPF map. To ensure that queries without a response are removed from the map, run a background garbage collection goroutine to remove timestamps older than a threshold.

## How to use

Follow the steps in [docs/gadgets/trace/dns.md](https://github.com/inspektor-gadget/inspektor-gadget/blob/main/docs/gadgets/trace/dns.md). To show latency, use `-o columns=latency`

## Testing done

1. Ran TestTraceDns integration test in minikube many times, verified that it passed consistently.
```
INTEGRATION_TESTS_PARAMS='-test.run TestTraceDns -no-deploy-ig -no-deploy-spo' make integration-tests
```

2. Manually tested in minikube with default namespace:
```
./kubectl-gadget trace dns -o columns=name,container,mntns,timestamp,id,qr,qtype,type,latency
```
<img width="937" alt="image" src="https://github.com/inspektor-gadget/inspektor-gadget/assets/2948394/076c0385-789a-41a5-9555-96814b5cc6c1">

3. Manually tested in minikube with all namespaces (includes host netns):
```
./kubectl-gadget trace dns -o columns=name,container,mntns,timestamp,id,qr,qtype,type,latency -A
```
<img width="941" alt="image" src="https://github.com/inspektor-gadget/inspektor-gadget/assets/2948394/ceb21364-c36f-4c3b-b3f4-d2e0311e90ad">

4. Increased log verbosity of garbage collector code I added and ran a DNS query to an IP without a DNS server (verify unresponded queries were GC'd):
<img width="927" alt="image" src="https://github.com/inspektor-gadget/inspektor-gadget/assets/2948394/bbbf4b2c-01b3-4e58-b88a-40fdc56812d1">

wedaly commented on pkg/gadgets/trace/dns/tracer/bpf/dns.c:
```diff
@@ -204,6 +221,35 @@ output_dns_event(struct __sk_buff *skb, union dnsflags flags, __u32 name_len, __
+	// to free space occupied by queries that never receive a response.
+	if (event->mount_ns_id) {
+		struct query_key_t query_key = {.mount_ns_id = event->mount_ns_id, .id = event->id};
+		if (event->qr == 0 && event->pkt_type == 0x4) { // query with type PACKET_OUTGOING
```
updated to use an `#define`. It looks like PACKET_HOST and PACKET_OUTGOING are also defined in [sockets-map.h](https://github.com/inspektor-gadget/inspektor-gadget/blob/03c29258d42e7d13126b7a83da8f89ce4e8d030e/pkg/gadgets/internal/socketenricher/bpf/sockets-map.h#L16-L22), so I put the `#define` in an `#ifndef` / `#endif` block.

wedaly replied:
thanks, added #define for query and response as well

wedaly commented on pkg/gadgets/trace/dns/tracer/gc.go:
```diff
@@ -0,0 +1,125 @@
+const garbageCollectorInterval = 1 * time.Second
+
+// Max age of query before it will be garbage collected.
+const garbageCollectorMaxQueryAge = 10 * time.Second
```
hmm, I don't have hard data here. I see that [`nslookup`](https://elixir.bootlin.com/busybox/latest/source/networking/nslookup.c#L331), [`dig`](https://linux.die.net/man/1/dig), and [coredns](https://github.com/coredns/coredns/blob/c1b479fd024c91d03608587765dc9ae8745aef0c/plugin/forward/forward.go#L259) all set a default timeout of 5 seconds, although a user could configure it higher.

Possibly we could make this user-configurable? Or set the default much higher (60 seconds?)

wedaly replied:
Updated to make query max age configurable (I called it "dns-timeout" but could change the name to whatever seems clearest).

I think I'll defer sending an event since that's tracked in a separate issue https://github.com/inspektor-gadget/inspektor-gadget/issues/1512

blanquicet reviewed (commented):
I have a few comments. I checked the general logic from the code and it seems correct. However, I haven't test it yet. We definitely need to add unit-test for this gadget to simulate different scenarios (not sure how complicated it might be).

wedaly commented on pkg/gadgets/trace/dns/tracer/bpf/dns-common.h:
```diff
@@ -42,6 +42,8 @@ struct event_t {
 	unsigned char pkt_type;
 	unsigned char rcode;
 
+	__u64 latency_ns; // Set only for responses in a container namespace with pkt_type HOST.
```
updated!

wedaly commented on pkg/gadgets/trace/dns/tracer/tracer.go:
```diff
@@ -326,12 +316,22 @@ func (t *Tracer) install() error {
+	}
+	t.gc = newGarbageCollector(queryMap)
+	t.gc.start()
+
```
moved to `Init`

wedaly commented on pkg/gadgets/trace/dns/tracer/gc.go:
```diff
@@ -0,0 +1,126 @@
+	queryMap *ebpf.Map
+}
+
+func newGarbageCollector(queryMap *ebpf.Map) *garbageCollector {
```
updated to use `gadgetCtx`. Also lowered the log level for some of these, since Info was too noisy when the logs show up in the console

wedaly commented on pkg/gadgets/trace/dns/tracer/gc.go:
```diff
@@ -0,0 +1,126 @@
+	"github.com/cilium/ebpf"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/gadgets"
+	"github.com/inspektor-gadget/inspektor-gadget/pkg/types"
+	log "github.com/sirupsen/logrus"
```
thanks, fixed!

wedaly commented on pkg/gadgets/trace/dns/tracer/gc.go:
```diff
@@ -0,0 +1,126 @@
+			log.Debugf("Executing DNS query map garbage collection")
+			gc.collect()
+			time.Sleep(garbageCollectorInterval)
+		}
```
would that allocate a channel on every loop? I guess we could use `time.Ticker` to create the channel once, but I wasn't sure if that had any advantages over `time.Sleep` here.

blanquicet replied:
You're right. Given that it will be executed multiple times, it is better to use the ticker. I believe the implementation is cleaner with a ticker than the sleep.

wedaly replied:
updated to use time.Ticker

wedaly commented on pkg/gadgets/trace/dns/tracer/tracer.go:
```diff
@@ -21,10 +21,9 @@ import (
 	"fmt"
 	"net/netip"
 	"syscall"
+	"time"
```
neat, I'll take a look at that. Do you think that should be part of this PR or a follow up?

blanquicet replied:
It would be great to have in this PR to test the changes we are introducing but it is fine in another PR so that we resolve the random fails.

blanquicet commented on pkg/gadgets/trace/dns/tracer/bpf/dns.c:
```diff
@@ -204,6 +232,36 @@ output_dns_event(struct __sk_buff *skb, union dnsflags flags, __u32 name_len, __
+	// Filter by mount_ns_id and pkt_type to track the latency from when a query is sent from a container
+	// and when a response to the query is received by that same container.
+	//
+	// * On DNS query sent from a container namespace (mount_ns_id != 0, qr == 0, and pkt_type == OUTGOING),
+	//   store the query timestamp in a map.
+	//
+	// * On DNS response received in the same container namespace (mount_ns_id != 0, qr == 1, and pkt_type == HOST)
+	//   retrieve/delete the query timestamp and set the latency field on the event.
```
```suggestion
	// * On DNS query sent from a container namespace (mount_ns_id != 0, qr == DNS_QR_QUERY, and pkt_type == OUTGOING),
	//   store the query timestamp in a map.
	//
	// * On DNS response received in the same container namespace (mount_ns_id != 0, qr == DNS_QR_RESP, and pkt_type == HOST)
	//   retrieve/delete the query timestamp and set the latency field on the event.
```

wedaly replied:
updated

blanquicet commented on pkg/gadgets/trace/dns/tracer/bpf/dns.c:
```diff
@@ -204,6 +232,36 @@ output_dns_event(struct __sk_buff *skb, union dnsflags flags, __u32 name_len, __
+	// to free space occupied by queries that never receive a response.
+	if (event->mount_ns_id) {
+		struct query_key_t query_key = {.mount_ns_id = event->mount_ns_id, .id = event->id};
+		if (event->qr == DNS_QR_QUERY && event->pkt_type == PACKET_OUTGOING) { // query with type PACKET_OUTGOING
+			struct query_ts_t query_ts = {.timestamp = event->timestamp};
+			bpf_map_update_elem(&query_map, &query_key, &query_ts, BPF_NOEXIST);
+		} else if (event->qr == DNS_QR_RESP && event->pkt_type == PACKET_HOST) { // response with type PACKET_HOST
```
nit: These two comments are not needed anymore.

wedaly replied:
thanks, removed them

blanquicet commented on pkg/gadgets/trace/dns/tracer/tracer.go:
```diff
@@ -280,7 +284,13 @@ func (t *Tracer) Init(gadgetCtx gadgets.GadgetContext) error {
 		return fmt.Errorf("installing tracer: %w", err)
 	}
 
+	if err := t.startGarbageCollector(gadgetCtx); err != nil {
+		t.Close()
+		return fmt.Errorf("starting garbage collector: %w", err)
+	}
+
 	t.ctx, t.cancel = gadgetcontext.WithTimeoutOrCancel(gadgetCtx.Context(), gadgetCtx.Timeout())
```
Actually, I think the `newGarbageCollector` should receive the `gadgetCtx.Logger()` and also the `t.ctx` to wait on that instead of having its own channel to stop GC'ing. That would be the same mechanism we have in [toppers](https://github.com/inspektor-gadget/inspektor-gadget/blob/c83014e7b57c32120c32edc0451e6c0490a396dc/pkg/gadgets/top/tcp/tracer/tracer.go#L228).

That's indeed the aim of having such a context: go-routines within gadgets will all get notified if the user stopped the gadget (Ctrl+C) or if it was executed with `--timeout` and such time has passed.

wedaly replied:
nice, that's simpler. Updated to use `t.ctx` Also changed it from a struct with methods to just a function `startGarbageCollector` since it no longer maintains state or needs a cleanup method.

blanquicet reviewed (approved):
LGTM! The CI is green, and I tested it, and it works as expected. However, I want someone with more experience reviewing the eBPF code before merging it.

wedaly commented on pkg/gadgets/trace/dns/tracer/gc.go:
```diff
@@ -0,0 +1,100 @@
+	}
+
+	for _, key := range keysToDelete {
+		logger.Debugf("deleting key with mntNs=%d and DNS ID=%x from query map for DNS tracer", key.MountNsId, key.Id)
```
changed it to a warning with the msg "DNS query with ID=%x in mntNs=%d did not receive a response within the DNS timeout (%s)"

blanquicet replied:
I don't know. If you're debugging the DNS traffic, I don't think it would be annoying to know that a given request didn't receive a response within X seconds (where X is btw configurable).

mauriciovasquezbernal replied:
I agree having a notification about a missing answer is good, but this message indicates something different, it's not that straightforward for the user to understand what's the real meaning of this.

wedaly replied:
I removed this log msg when switching from the map iterator to batch lookup/delete. Will defer adding an event/notification since that's tracked by issue https://github.com/inspektor-gadget/inspektor-gadget/issues/1512

blanquicet commented:
BTW, please squash the commits before merging. They can all go in a single commit, IMO.

mauriciovasquezbernal reviewed (commented):
Hey @wedaly, thanks for this contribution. 

This latency calculation only works on 5.8+ because it uses [bpf_ktime_get_boot_ns](https://github.com/torvalds/linux/commit/71d19214776e61b33da48f7c1b46e522c7f78221) (same applies to previous solution). I think we need to put all the logic inside a conditional in the eBPF code to avoid doing useless work (like saving in the map) when that field can't be calculated. DetectBpfKtimeGetBootNs() can be used to check if the needed helper is available.

Regarding the cleanup part, we can use an [ebpf iterator on the map](https://github.com/torvalds/linux/commit/a5cbe05a6673b85bed2a63ffcfea6a96c6410cff), [as done in the socket enricher](https://github.com/inspektor-gadget/inspektor-gadget/blob/03c29258d42e7d13126b7a83da8f89ce4e8d030e/pkg/gadgets/internal/socketenricher/bpf/sockets-iter.bpf.c#L98), that's more efficient as we don't need to perform a lot of system calls. The only issue is that map iterators are only available from 5.9+, so we have few options:
1. Support 5.8+ with the cleanup implemented in user space
2. Support 5.9+ with the cleanup implemented using bpf iterators
3. Support 5.8+ with both cleanups implemented, user space only when bpf iterators aren't available.

Either solution is fine to me, but please add a warning that tells the user when latency is not available.

wedaly commented on pkg/gadgets/trace/dns/tracer/gc.go:
```diff
@@ -0,0 +1,109 @@
+	return nil
+}
+
+func collectGarbage(logger logger.Logger, dnsTimeout time.Duration, queryMap *ebpf.Map) {
```
Nice, I didn't realize these were available. Updated to use `BatchLookup` and `BatchDelete`. I set the batch size to 256 so worst-case this would require 1024 / 256 = 4 lookup and delete calls, but could adjust that higher or lower. Tested it by running 5000 nslookup calls to a bad IP in parallel and verified from the logs that 1024 entries were collected.

wedaly commented on pkg/gadgets/trace/dns/tracer/bpf/dns.c:
```diff
@@ -88,6 +99,23 @@ struct {
+struct query_key_t {
+	__u64 mount_ns_id;
+	__u16 id;
+};
```
updated to use pid_tgid as the map key

wedaly commented on pkg/gadgets/trace/dns/tracer/bpf/dns.c:
```diff
@@ -204,6 +232,36 @@ output_dns_event(struct __sk_buff *skb, union dnsflags flags, __u32 name_len, __
+	//
+	// A garbage collection thread running in userspace periodically scans for keys with old timestamps
+	// to free space occupied by queries that never receive a response.
+	if (event->mount_ns_id) {
```
removed the check for mount_ns_id

wedaly commented:
> This latency calculation only works on 5.8+ because it uses [bpf_ktime_get_boot_ns](https://github.com/torvalds/linux/commit/71d19214776e61b33da48f7c1b46e522c7f78221) (same applies to previous solution). I think we need to put all the logic inside a conditional in the eBPF code to avoid doing useless work (like saving in the map) when that field can't be calculated. DetectBpfKtimeGetBootNs() can be used to check if the needed helper is available.

Ah, I forgot about that. Added a warning if `!DetectBpfKtimeGetBootNs` in `gc.go` and returned early so we never start the GC goroutine.

For skipping the map update in the BPF program, I added a check that `event->timestamp> 0` (on the assumption that the stub `bpf_ktime_get_boot_ns` injected by [`removeBpfKtimeGetBootNs`](https://github.com/inspektor-gadget/inspektor-gadget/blob/b57f2bae31a46b40d8e0204b85099ae37f15d21d/pkg/gadgets/helpers.go#L249-L252) should return zero. Is there a better way to detect this?

> Regarding the cleanup part, we can use an [ebpf iterator on the map](https://github.com/torvalds/linux/commit/a5cbe05a6673b85bed2a63ffcfea6a96c6410cff), [as done in the socket enricher](https://github.com/inspektor-gadget/inspektor-gadget/blob/03c29258d42e7d13126b7a83da8f89ce4e8d030e/pkg/gadgets/internal/socketenricher/bpf/sockets-iter.bpf.c#L98), that's more efficient as we don't need to perform a lot of system calls. The only issue is that map iterators are only available from 5.9+, so we have few options:
> 
> 1. Support 5.8+ with the cleanup implemented in user space
> 2. Support 5.9+ with the cleanup implemented using bpf iterators
> 3. Support 5.8+ with both cleanups implemented, user space only when bpf iterators aren't available.
> 
> Either solution is fine to me, but please add a warning that tells the user when latency is not available.

Thanks for the pointer! I decided to keep the GC in userspace since I think it will make it easier to implement https://github.com/inspektor-gadget/inspektor-gadget/issues/1512 if we wanted to report missing answers. With batch lookup/delete, I think the number of syscalls should be reasonable even when the map is full.

mauriciovasquezbernal commented on pkg/gadgets/trace/dns/tracer/bpf/dns.c:
```diff
@@ -204,6 +232,43 @@ output_dns_event(struct __sk_buff *skb, union dnsflags flags, __u32 name_len, __
 	int anaddrcount = load_addresses(skb, ancount, anoffset, event);
 	event->anaddrcount = anaddrcount;
 
+	// Calculate latency:
+	//
+	// Filter by mount_ns_id and pkt_type to track the latency from when a query is sent from a container
+	// and when a response to the query is received by that same container.
+	//
+	// * On DNS query sent from a container namespace (mount_ns_id != 0, qr == DNS_QR_QUERY, and pkt_type == OUTGOING),
+	//   store the query timestamp in a map.
+	//
+	// * On DNS response received in the same container namespace (mount_ns_id != 0, qr == DNS_QR_RESP, and pkt_type == HOST)
```
Please update this comment as `mount_ns_id` isn't used anymore.

mauriciovasquezbernal commented on pkg/gadgets/trace/dns/tracer/bpf/dns.c:
```diff
@@ -204,6 +232,43 @@ output_dns_event(struct __sk_buff *skb, union dnsflags flags, __u32 name_len, __
+	// Skip this if skb_val == NULL (gadget_socket_lookup did not set pid_tgid we use in the query key)
+	// or if event->timestamp == 0 (kernels before 5.8 don't support bpf_ktime_get_boot_ns, and the patched
+	// version IG injects always returns zero).
+	if (skb_val != NULL && event->timestamp > 0) {
+		struct query_key_t query_key = {
+			.pid_tgid = skb_val->pid_tgid,
```
What do you think about removing the `skb_val != NULL` check and just use 0 in case it isn't available?

```suggestion
	if (event->timestamp > 0) {
		struct query_key_t query_key = {
			.pid_tgid = skb_val ? skb_val->pid_tgid : 0,
```

IMO that's a good fallback when the socket enricher it's not available. @alban  do you have comments on this?

wedaly replied:
Would that get confused with `trace dns -A`? I'm thinking about what happens to a DNS packet that goes from client -> coredns -> upstream DNS and back again. I think the DNS ID is the same when coredns forwards the query upstream, and if we're setting pid_tgid to zero then query_key would also be the same.

If all the containers are using the same query_key, then I think latency would be reported for the response back to coredns, but calculated from when the request was sent from the client (not when it was sent from coredns)

wedaly commented on pkg/gadgets/trace/dns/tracer/bpf/dns.c:
```diff
@@ -88,6 +99,23 @@ struct {
 
+// Map of DNS query to timestamp so we can calculate latency from query sent to answer received.
+struct query_key_t {
+	__u32 pid_tgid;
```
ooh, good catch, thanks! updated

wedaly commented on pkg/gadgets/trace/dns/tracer/gc.go:
```diff
@@ -0,0 +1,131 @@
+	dnsTimeout := gadgetParams.Get(ParamDNSTimeout).AsDuration()
+	if dnsTimeout <= 0 {
+		return fmt.Errorf("DNS timeout must be > 0")
+	}
```
nice, didn't know that. Moved the validation into ParamDesc.Validator
